docs/architecture.md

```diff
@@ -0,0 +1,292 @@
+- `spark.armada.Driver.labels` - A comma-separated list of kubernetes labels
+    (in key=value format) to be added to the Driver pod.
+- `spark.armada.Executor.labels` A comma-separated list of kubernetes labels
+    (in key=value format) to be added to all Executor pods.
+
+# Building `armada-spark`
+
+## Scala Source
+
+`armada-spark` can be built from source using maven. See the [Building Armada Spark](https://github.com/armadaproject/armada-spark/blob/master/README.md#building-armada-spark)
+section of the README for more information.
+
+## Docker image
+
+`armada-spark` uses docker to produce container images that serve as spark
+Drivers and Executors. See the [Building Docker Images](https://github.com/ClifHouck/armada-spark/blob/master/README.md#building-docker-images)
+section of the README for more information.
+
+# Architecture & Design
+
+## Overview
+
+Fundamentally, `armada-spark`'s architecture and design is closely related to
+the existing Spark Kubernetes cluster manager, but with the added layer of
+Armada and its job system on top.
+
+A great overview of what a "cluster" is from Spark's perspective can be found
+[here](https://spark.apache.org/docs/latest/cluster-overview.html).
+
+```
+// TODO: armada-spark Architecture overview diagram.
+```
+
+The basic flow of an `armada-spark` execution is as follows:
+
+1. End-user calls `ArmadaSparkSubmit.main` via `spark-submit` or `spark-class`.
+2. `ArmadaSparkSubmit`:
+    - Constructs an `ArmadaClient` to communicate with Armada Server via gRPC.
+    - Verifies connectivity and service health of specified Armada URL.
+    - Submits a Driver job to Armada.
+    - Submits `n` Executor jobs to Armada.
+3. After job submissions, `ArmadaSparkSubmit` prints a Lookout link and exits.
+4. Armada schedules Driver and Executor jobs submitted by `ArmadaSparkSubmit`.
+    - Each job submits a pod with a single container.
+    - By using a common label, we can guarantee Driver and Executor jobs are
+    scheduled to the same cluster. This guarantees network connectivity and has
+    the added benefit of ensuring Drivers and Executors run relatively close to
+    each other, reducing network latency and overhead.
+5. Asynchronously:
+    - Driver starts up and waits for each expected Executor to establish
+    connections.
+    - Each Executor starts and establishes connections to the Driver.
+6. Driver and Executors proceed as any normal Spark program would.
+7. Driver and Executors finish their work.
+
+## Integrating with Spark, or "What is a Cluster Manager?"
+
+A "Cluster Manager" is a concept within Spark which allows Spark to submit and
+manage Drivers and Executors to a compute cluster like YARN, Mesos,
+Kubernetes, and now Armada. Confusingly, cluster managers are also sometimes
+referred to as "resource-managers" within the Spark codebase.
+
+### Defining a new Cluster Manager
+
+To define a new cluster manager within Spark one would inherit the `trait`
```

Comment:
And change the perspective here, something like:
The AS CM is defined in Spark by inheriting ...

Reply:
Addressed in 7c56abb